CefSharp.Wpf(Chromium) を使った艦これ 第二期への対応

Mine02C4 opened the pull request:
## 目的

艦これ 第二期(HTML5)【Block-1】への対応

## 内容

技術実証的なレベルだが、 CefSharp.Wpf を用いて対応させた。ひとまず、Nekoxyに通信刺して、基本的な解析機能が動くのは確認したた。通信内容はあまり変わっていない模様。

![kcap](https://user-images.githubusercontent.com/304782/44295263-712dc880-a2e0-11e8-8f7c-45121f2f5b79.png)

CefSharp.Wpf は Chromium を .NET に埋め込むライブラリのWPF版である。単純にWebBrowserをこれに置き換えた。Edgeを使うアプローチを調べてるようだったので、こっちを試してみた。

Debugモードでビルドすると http://127.0.0.1:28088 にChromeでアクセスすることで開発者ツールが使える。(たまに固まる…)

## 制限やまだやってないこと

### 1. KanColleViewerをx64(or x86)でビルドする必要がある。

Chromium自体は Unmanaged なので、AnyCPUではビルドできない。このPRではひとまずx64をターゲットにしている。設定次第でx86も可能のはずです。KanColleViewerだけで良いので、ほかはAnyCPUのまま。

### ~~2. ゲーム画面の調整をまだしていない~~

~~まだスクロールバーが出っぱなしのまま。CSSを入れる方法はまだ調査中。~~
JavaScriptでstyleタグを追加。(2重追加のロジック入れてないので後で入れる)

### ~~3. スクリーンショットはまだ~~

~~2がまだなので~~

WebGLを無効化することで、canvasの描画結果を保存

### 4. HiDPI対応

ズームだけでなく、それらしいプロパティが CefSharp にあるので、よしなに対応できそう(要確認)

### 5. URL遷移時の挙動がキモい(WPF上で不可視になる)

設定ですぐ解決できるかもしれないが、現状ではURL遷移時の読み込み中にWPFの要素が見えなくなる。(白背景とかではなく、その後ろのグレーが見える)

Mine02C4 committed:
Add CefSharp (and make KanColleViewer target x64

Mine02C4 committed:
Integrate NavigatorBehavior

Mine02C4 committed:
Setting CefSharp Proxy(But configuration is hard-corded

Mine02C4 committed:
Compatible with 2nd season

Mine02C4 committed:
Plugins関連に対応するため OutputPath を暫定的に変更

WolfgangKurz commented:
2.と3.は次のURLを参考してください。
今はEdgeブラウザー向けのシステムに変更中なので使ってはいませんが…

https://github.com/CirnoV/KanColleViewer/tree/611079d76aa5d5f42dce5cf2c181280036dea921

CSSはJSで<style>タグを生成、スクショはChromiumWebBrowserクラス内部のビットマップメモリをコピーして実装しました。

KanColleHost.csとScreenshotAction.csとModels/Cef/ChromiumWebBrowserEx.csだけが重要なので、それだけ見ても結構です。

Mine02C4 commented:
ありがとうございます。
そのままマージはできなそうですので、参考にしつつ移植してみます。

Mine02C4 committed:
ゲーム画面のサイズ調整関連ロジックの追加

Ported from https://github.com/CirnoV/KanColleViewer/commit/feb0395044c3a5d3f299b5bd3f286334b1864c4a

Mine02C4 committed:
SolutionレベルでのAnyCPUを削除

Mine02C4 committed:
Frameを取得してtoDataURLを試みるが真っ黒

WebGLを使っているのが原因と見られる。

Mine02C4 committed:
WebGLを無効化することで、固定解像度でスクリーンショットを取得

Mine02C4 committed:
boolは0,1で指定するらしい

Mine02C4 commented:
不要になったので閉じておきます。
